### ISSUE TITLE
Multiplication of `torch.tensor` with `np.array` does the operation with numpy.

### ISSUE THREAD

**Issue by boeddeker** [pytorch/pytorch#59237]
## 🐛 Bug

 - Bug: Multiplication of `torch.tensor` with `np.array` silently does the operation with numpy.
 - Correct: Multiplication of `np.array` with `torch.tensor` fails as expected.
 - Insight: Multiplication of `torch.tensor(requires_grad=True)` with `np.array` indicates that the op is done with numpy.

## To Reproduce

The code to show the effect:
```python
import numpy as np
import torch

print(np.__version__)
print(torch.__version__)

a = np.arange(10).astype(np.float32)
a_pt = torch.tensor(a, requires_grad=False)
a_pt * a
# a * a_pt  # -> TypeError: unsupported operand type(s) for *: 'numpy.ndarray' and 'Tensor'
a_pt = torch.tensor(a, requires_grad=True)
a_pt * a
```
and the output
```python
1.19.2
1.7.1+cpu
---------------------------------------------------------------------
RuntimeError                        Traceback (most recent call last)
<ipython-input-26-f0b1e2a27462> in <module>
     10 # a * a_pt  # -> TypeError: unsupported operand type(s) for *: 'numpy.ndarray' and 'Tensor'
     11 a_pt = torch.tensor(a, requires_grad=True)
---> 12 a_pt * a

/opt/conda36/lib/python3.8/site-packages/torch/tensor.py in __array__(self, dtype)
    628             return handle_torch_function(Tensor.__array__, relevant_args, self, dtype=dtype)
    629         if dtype is None:
--> 630             return self.numpy()
    631         else:
    632             return self.numpy().astype(dtype, copy=False)

RuntimeError: Can't call numpy() on Tensor that requires grad. Use tensor.detach().numpy() instead.

```

## Expected behavior

Either fail or do the operation with torch.

## Environment

```
PyTorch version: 1.7.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.14.0
Libc version: glibc-2.31

Python version: 3.8 (64-bit runtime)
Python platform: Linux-5.4.0-73-generic-x86_64-with-glibc2.10
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] numpydoc==0.9.2
[pip3] padertorch==0.0.0
[pip3] torch==1.7.1+cpu
[pip3] torch-complex==0.1.2
[pip3] torchaudio==0.8.0.dev20210222
[pip3] torchvision==0.9.0.dev20210222+cpu
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.1                      217  
[conda] mkl-include               2020.1                      217  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.2.1            py38h54f3939_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.19.2           py38h54aff64_0  
[conda] numpy-base                1.19.2           py38hfa32c7d_0  
[conda] numpydoc                  0.9.2                    pypi_0    pypi
[conda] padertorch                0.0.0                     dev_0    <develop>
[conda] torch                     1.7.1+cpu                pypi_0    pypi
[conda] torch-complex             0.1.2                     dev_0    <develop>
[conda] torchaudio                0.8.0.dev20210222          pypi_0    pypi
[conda] torchvision               0.9.0.dev20210222+cpu          pypi_0    pypi
```

## Additional context

Sorry, if this is a duplicate, but I was not able to find if someone already reported this.
I only found discussions about the `__array__` function.

**Comment by ngimel** [comment]
Tentatively hi pri for unexpected behavior

**Comment by rgommers** [comment]
> Either fail or do the operation with torch.

This is not the expected behavior. PyTorch does not work on `numpy.ndarray` instances, while vice versa NumPy _does_ accept `torch.Tensor` instances. To show that:
```python
>>> import numpy as np
>>> import torch
>>> a = np.arange(10).astype(np.float32)
>>> a_pt = torch.tensor(a, requires_grad=False)
>>> np.multiply(a, a_pt)
tensor([ 0.,  1.,  4.,  9., 16., 25., 36., 49., 64., 81.])
>>> np.multiply(a_pt, a)
tensor([ 0.,  1.,  4.,  9., 16., 25., 36., 49., 64., 81.])
>>> torch.multiply(a, a_pt)
...
TypeError: multiply() received an invalid combination of arguments - got (numpy.ndarray, Tensor), but expected one of:
 * (Tensor input, Tensor other, *, Tensor out)
 * (Tensor input, Number other)
      didn't match because some of the arguments have invalid types: (numpy.ndarray, Tensor)

>>> torch.multiply(a_pt, a)
...
TypeError: multiply() received an invalid combination of arguments - got (Tensor, numpy.ndarray), but expected one of:
 * (Tensor input, Tensor other, *, Tensor out)
 * (Tensor input, Number other)
      didn't match because some of the arguments have invalid types: (Tensor, numpy.ndarray)
```

Now for the `*` operator, it depends on the order:
```python
>>> a_pt * a  # note: this does the multiplication with numpy, then returns a torch.Tensor!
tensor([ 0.,  1.,  4.,  9., 16., 25., 36., 49., 64., 81.])
>>> a * a_pt
...
TypeError: unsupported operand type(s) for *: 'numpy.ndarray' and 'Tensor'
```

The root cause of that is:
```python
>>> a_pt.__mul__(a)
NotImplemented
>>> a_pt.__rmul__(a)
NotImplemented
>>> a.__mul__(a_pt)
NotImplemented
>>> a.__rmul__(a_pt)
tensor([ 0.,  1.,  4.,  9., 16., 25., 36., 49., 64., 81.])
```

The problematic case here is `a * a_pt`, which doesn't work because `a.__mul__(a_pt)` returns `NotImplemented`. The resolution methods for `__op__`/`__rop__` in NumPy are pretty complicated (and depend on the presence/absence of some `__Tensor.__array_*__` methods), and I'm not sure if this is actually supposed to work or not. 

It's not just for `*`, something different happens for each operator:

```python
>>> a.__radd__(a_pt)
tensor([ 0.,  2.,  4.,  6.,  8., 10., 12., 14., 16., 18.])
>>> a.__add__(a_pt)
NotImplemented
>>> a + a_pt
...
    a + a_pt
TypeError: Concatenation operation is not implemented for NumPy arrays, use np.concatenate() instead. Please do not rely on this error; it may not be given on all Python implementations.

>>> a_pt + a
tensor([ 0.,  2.,  4.,  6.,  8., 10., 12., 14., 16., 18.])


>>> a_pt / a
tensor([nan, 1., 1., 1., 1., 1., 1., 1., 1., 1.])
>>> a / a_pt
tensor([nan, 1., 1., 1., 1., 1., 1., 1., 1., 1.])

>>> a_pt - a
tensor([0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])
>>> a - a_pt
...
TypeError: rsub() received an invalid combination of arguments - got (Tensor, numpy.ndarray), but expected one of:
 * (Tensor input, Tensor other, *, Number alpha)
 * (Tensor input, Number other, Number alpha)
```

So `-` is like `*`, while `/` works and `+` somehow hits concatenate.

tl;dr this is a bit of a mess in NumPy. PyTorch just returns `NotImplemented` and works as expected.

**Comment by boeddeker** [comment]
Yes, I agree, that it is better not to rely on this. Especially since it does not work with GPU Tensors (assume, since `cpu` is missing in `self.numpy()` in `def __array__`) and also does not work, when a grad is necessary.
I accidentally hit this in my code, where I tested everything with CPU and without gradients.
Later I tried to use it with gradients and I got this error.

When I understand https://numpy.org/neps/nep-0030-duck-array-protocol.html correctly,
NumPy intends, that `np.multiply` uses torch functions, when both arguments are torch.

So relying on an implicit conversion may break future ideas, where someone writes a function 
with numpy that can be used with different backends.

**Comment by rgommers** [comment]
> When I understand https://numpy.org/neps/nep-0030-duck-array-protocol.html correctly,
> NumPy intends, that `np.multiply` uses torch functions, when both arguments are torch.

NEP 30 isn't relevant here, since PyTorch doesn't (and likely will never) support `__array_function__`. When you call `np.multiply(a, b)`, then `a` and `b` get converted to `ndarray` - it's just more explicit than relying on operator precedence for `*`.

> So relying on an implicit conversion may break future ideas, where someone writes a function
with numpy that can be used with different backends.

indeed
